### ISSUE TITLE
[ON HOLD][$2000] Android - Workspace is still selectable even after 'Create room' button has been pressed

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#21577]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to android
2. Click on FAB menu
3. Click on New Room
4. Enter room name > select a workspace
5. (Important step )> Click on create room and immediately again select the workspace dropdown and notice that it is still clickable

## Expected Result:
Workspace should not be selected when the 'create room' button has been pressed

## Actual Result:
Workspace is still selectable even after 'Create room' button has been pressed

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.29-11
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/9d5fd6ae-444f-4d53-ae03-dcf7a01e3060

https://github.com/Expensify/App/assets/93399543/d299d0d5-70ce-4586-ab4b-8fbb36f34c8f

**Expensify/Expensify Issue URL:**
**Issue reported by:** @priya-zha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1687430588059259

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011eacb8f4d46da0d6</li>
        <li>Upwork Job ID: 1673652478476353536</li>
        <li>Last Price Increase: 2023-07-19</li>
    </ul>
</details>

**Comment by Thanos30** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Picker (Selection fields) are still enabled after the Create Room button is clicked, which allows the user to change the values.

### What is the root cause of that problem?

The root cause of this problem is on the `WorkspaceNewRoomPage.js` file:

```
submit(values) {
        const policyMembers = _.map(_.keys(this.props.allPolicyMembers[`${ONYXKEYS.COLLECTION.POLICY_MEMBERS}${values.policyID}`]), (accountID) => Number(accountID));
        Report.addPolicyReport(values.policyID, values.roomName, values.visibility, policyMembers);
    }
```

When the submit function is called, this doesn't block the Pickers in any way, so while the action is fired, they are still enabled.

### What changes do you think we should make in order to solve the problem?

We should add an extra state on the component like that:
```
constructor(props) {
        super(props);

        this.state = {
            visibilityDescription: this.props.translate('newRoomPage.restrictedDescription'),
            isFormSubmitting: false, // New property for form submission state

        };

```

On the submit action, we should add the following:

```
submit(values) {
        const policyMembers = _.map(_.keys(this.props.allPolicyMembers[`${ONYXKEYS.COLLECTION.POLICY_MEMBERS}${values.policyID}`]), (accountID) => Number(accountID));
        this.setState({isFormSubmitting: true});
        Report.addPolicyReport(values.policyID, values.roomName, values.visibility, policyMembers);
        this.setState({isFormSubmitting: false});

    }
```

Finally, we will add the isDisabled prop on the Pickers:

```
<Picker
                            inputID="policyID"
                            label={this.props.translate('workspace.common.workspace')}
                            placeholder={{value: '', label: this.props.translate('newRoomPage.selectAWorkspace')}}
                            items={workspaceOptions}
                            isDisabled={this.state.isFormSubmitting}
                        />
```

This way, the selection fields / inputs will be disabled while the room is being created

**Comment by johncschuster** [comment]
Not overdue, Melvin. We're just getting proposals now.

@fedirjh, what do you think of the above proposal?

**Comment by Thanos30** [comment]
> We should fix it globally, it may occur on other forms as well, so it would be ideal to address the root cause.

I checked, I can follow the same logic inside the Form component so that it works like that everywhere.

> Maybe we should investigate that further, this is feels tricky to me. Not sure , but we can switch the order of execution, update state then submit the form.

I tried that, as I mentioned above, the state update is asynchronous, so it still doesn't trigger. I understand that it sounds a bit tricky, in real life scenarios the Navigation will probably never take long enough for the user to be able to click on the selection fields again.

Debouncing the creation of the room and the navigation with a delay of 0 will simply run the actions on the background asynchronously, which allows us to remain on the same state and be able to update it. If you think about it, in most forms, there is an asynchronous function following up, so it's not that strange.

**Comment by johncschuster** [comment]
@fedirjh what are your thoughts on @Thanos30's comment above?

**Comment by johncschuster** [comment]
Not overdue, Melvin. We're still waiting on proposals

**Comment by Thanos30** [comment]
@fedirjh I believe that handling the navigation asynchronously (using the debounce) instead of synchronously is a good solution, as it won't block the state update. Do you disagree with that?

**Comment by kmwamasali** [comment]
@fedirjh do I need a high traffic account to test this. I am unable to access the new room option on my FAB menu.
Kindly advise on steps needed to replicate

**Comment by kmwamasali** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Workspace is still selectable even after 'Create room' button has been pressed

### What is the root cause of that problem?
Form elements can still be clicked during the submission event, with only the submit button being disabled to avoid duplicate submissions.
https://github.com/Expensify/App/blob/b9508b616f3a7229365c8bd7280fd1e34cfd2f9b/src/components/Form.js#L177-L181

### What changes do you think we should make in order to solve the problem?
By wrapping all form elements within `<fieldset disabled>` all inputs can be blocked during submission
see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset#disabled_fieldset
We can use the already existing `isLoading` prop to handle enabling/diabling the input fields
https://github.com/Expensify/App/blob/b9508b616f3a7229365c8bd7280fd1e34cfd2f9b/src/components/Form.js#L40-L43

### What alternative solutions did you explore? (Optional)
Because introducing a higher level DOM element in the form may or maynot (pending testing) introduce some issues from event bubbling concerning the input elements.
We can optionally pass a `disabled` attribute to the input elements which is also dependent on the form `isLoading` prop to handle enabling/diabling. By passing the field to the `childrenWrapperWithProps` which iterates over all input elements in the form
https://github.com/Expensify/App/blob/b9508b616f3a7229365c8bd7280fd1e34cfd2f9b/src/components/Form.js#L195-L204

**Comment by Thanos30** [comment]
@fedirjh If I may add here, I researched this a bit, and it is not bad practice to handle the navigation asynchronously.

The reason we are using `setTimeout` is because otherwise, the navigation blocks the state update, which blocks us from blocking the form fields. I explained it in more details above.

Thank you for the feedback 🙏

**Comment by kmwamasali** [comment]
@fedirjh yes, it would require a npm pakcakage to work. 

The input fields can still be passed the disabled field as I set in the alternative proposal. And by using the same loading state that blocks a second submission the input fields would also be disabled directly

**Comment by johncschuster** [comment]
@fedirjh what do you think of the above proposals?

**Comment by kmwamasali** [comment]
@fedirjh that makes sense

**Comment by kmwamasali** [comment]
@fedirjh I had tried using the default isDisabled prop on the `Picker` but because all form elements are touched during validation, it still deactivates just before submission which is too late and the user experience is as documented in the bug. Will attempt to work on the `Form` as suggested for an appropriate proposal. Thank you for the guidance

**Comment by johncschuster** [comment]
@fedirjh bump! Do you feel the proposals are sufficient, or should we take this internal?

**Comment by kmwamasali** [comment]
@fedirjh I managed to stop clicks using the css `pointer-events: none` but it required duplicating a huge piece of the form code block to introduce the style conditionally(easier suggestions to do this very welcome) . My attempts at using `preventDefault` to stop event bubbling didn't work.
My proposal would be to introduce a loading screen/overlay as it would also improve user experience since it's kind of akward to stare and wait after you've pressed the button with no indication that something is happening ?

**Comment by pasyukevich** [comment]
Hi,
I am from Callstack - an expert contributor group
I start to investigate this issue

**Comment by kmwamasali** [comment]
@fedirjh do you think a loading screen/mask for forms in general would be an okay idea

**Comment by pasyukevich** [comment]
I continue my further investigation by looking for a proper fix.

**Comment by Thanos30** [comment]
@fedirjh Could you please explain to me how Onyx is blocking the form state update?

I believe the navigation is what's blocking the state update, and that's why I handle the navigation asynchronously in my proposal.

Thanks 🙏

**Comment by Thanos30** [comment]
Thank you for the context @fedirjh. In my proposal, I am handling the whole creating the workspace method, including Onyx.SET and the navigation, asynchronously, which I still believe is the best case scenario here. While I was checking the delay though (debugging the issue) I noticed that the navigation action was fired, and the delay was on that part and not the SET on Onyx. I might be wrong though.

Is there any reason why you don't agree with this? I believe in such cases we could use asynchronous code, and if I am not mistaken is good practice too.

Thank you for your feedback once again 🙏

**Comment by johncschuster** [comment]
@fedirjh given we're at the WAQ threshold, what do you think we should do to get this over the line?

**Comment by pasyukevich** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
Input fields with selection are clickable after submission of the form
### What is the root cause of that problem?
`onSubmit` function consists of blocking thread code and delaying the transition to the next string. While the thread is blocked we are still able to interact with pickers. To fix this we need to block the interaction with the form before calling the submit function.
### What changes do you think we should make in order to solve the problem?
Add `isFormSubmitted` state and set it to true at the submit function
Wrap form content to block interaction based on `isFormSubmitted` state
```
<View style={{pointerEvents: isFormSubmitted ? 'none':'auto'}}>
        {childrenWrapperWithProps(_.isFunction(children) ? children({inputValues}) : children)}
        </View>
```
Add `useEffect` that will be triggered on `isFormSubmitted` change.
To establish that `onSubmit` function will non-blocking for repaint on the state changes we need to put `onSubmit` in the `requestAnimationFrame`, which will execute callback at the optimal time for smooth performance. (Callback and promise are not achieving the expected result in this case)
```
useEffect(() => {
      if (!isFormSubmitted) {
        return;
      }

     requestAnimationFrame(() => {
      onSubmit(inputValues);
     });
      
  }, [onSubmit, inputValues, isFormSubmitted]);
```
Update submit button loading state with `isFormSubmitted`
`isLoading={props.formState.isLoading || isFormSubmitted}`
### What alternative solutions did you explore?
`setTimeout` can be here as a working option but is not preferable as it does not guarantee that the callback will be executed at the specified time in sync with the screen refresh rate

**Comment by Thanos30** [comment]
@pasyukevich Isn't this basically the same solution that I proposed? The only 2 differences is:

1) The use of `requestAnimationFrame` instead of `setTimeout`, which, based on [here](https://reactnative.dev/docs/timers#timers), I believe it's better to use `setTimeout` and trigger the creation of the workspace and the navigation immediately.

2) The `useEffect` hook instead of altering the onSubmit callback. Isn't it much better to alter the onSubmit callback to use `requestAnimationFrame` (as you did in your proposal) on that specific code block that causes the issue, instead of making a `useEffect` hook that will be triggered on every change we do on the workspace creation page?

**Comment by pasyukevich** [comment]
@Thanos30 
  
  1. According to the [docs](https://reactnative.dev/docs/timers#timers) `requestAnimationFrame` will be called after flushing the frames. In this case, this guarantee that fields will be blocked from interactions 
  2. Here `useEffect` is used to schedule `onSubmit` only after the `isFormSubmitted` is updated. Also, I believe that better to cover this case for other forms too.

**Comment by Thanos30** [comment]
@pasyukevich Gotcha, thanks for the explanation 🙏

If the submission action, and therefore the navigation runs in `requestAnimationFrame` or `setTimeout` , it's gonna go on the background and run on the specified time, and only after the stack is cleared, and in this case, it's not ever gonna block the state update, so the faster the better, even if we are talking about miliseconds here.

Since this is happening on the background, there is no need to submit the form with a state update either, as these will run the moment the code is finished updating the state.

Also, not sure how handy it is to have a hook that will re-trigger on every input change, even if it's just returning nothing. This will happen on the Form component, so it will happen on all the forms throughout the app.

I liked the thought though!

**Comment by johncschuster** [comment]
@fedirjh now that we've received an update from @pasyukevich, what's the next step here?

**Comment by johncschuster** [comment]
Bumped in Slack.

**Comment by pasyukevich** [comment]
@fedirjh Thanks for review. I will update according to your comments

Can you share more about cases when we will receive an error on form submission? (I cannot find any resources or code that handles it)
As far as I know, we are using an optimistic approach.

**Comment by Thanos30** [comment]
@fedirjh Correct me if I am wrong please, but when I posted the first proposal here, your only concern was the use of `setTimeout` which you were not fond of. My proposal was simply running the submit action in Form using `setTimeout` which solved the issue. Is there something I am missing here?

Kinda feeling a bit aggrieved here since I was here 2 months ago with a proposal that worked and was trying to figure out a way to not use `setTimeout` or its alternatives as you suggested, but it seems that it is acceptable now.

Thanks in advance 🙏

**Comment by pasyukevich** [comment]
I will post an update here today

**Comment by pasyukevich** [comment]
Is it still an issue? 

While checking another solution I found that with current main delay is become small and is now the same as on the iOS platform.

**Comment by johncschuster** [comment]
@pasyukevich I'm having a hard time understanding what you mean in your comment above. Are you saying this behavior _is_ still reproducible, or is it no longer reproducible?

**Comment by pasyukevich** [comment]
@johncschuster Before, I was able to reproduce this issue with a long delay, which is similar to that in the reported video. We have the point that this issue was exclusive to Android due to Onyx's blocking behavior. (https://github.com/Expensify/App/issues/21577#issuecomment-1665877050)

Currently, this delay has significantly reduced and is consistent across platforms, including iOS. During this brief delay, between form submission and redirection to the next screen, users can still interact with the form fields on both iOS and Android.

Should we address this on both platforms or considering the minimal delay, should we not categorize it as a bug?

**Comment by johncschuster** [comment]
I think we can categorize this as `not a bug`. I believe we have a separate project going to address overall performance that will likely resolve the delay.

**Comment by johncschuster** [comment]
Thanks for the bump! That sounds good to me!

**Comment by johncschuster** [comment]
I've put the issue on hold. @fedirjh, can you confirm I have the right issue for the hold when you get a moment? Thank you!

**Comment by johncschuster** [comment]
Still holding.

**Comment by kbecciv** [comment]
@johncschuster Checking, will update you shortly

**Comment by kbecciv** [comment]
Issue is not reproducible

https://github.com/Expensify/App/assets/93399543/441bff17-64a9-4f62-a5b3-f18ea78b40f3

**Comment by johncschuster** [comment]
Yes! Let's close it!
